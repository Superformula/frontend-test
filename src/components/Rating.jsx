import React from 'react';
import PropTypes from 'prop-types';
import StarRatings from 'react-star-ratings';
import {
    ratingContainer,
    filledStarsContainer,
    emptyStarsContainer,
    emptyStar,
    starsContainer
} from './Rating.module.scss';

/*
The react-star-ratings module takes care of filling stars based on rating numbers, and can partially fill them etc.
However it doesn't have any way to make its "empty" stars just have a stroke around them - they can only be a solid color.
So we roll our own empty stars, swiping the svg path from the one react-star-ratings uses.
*/
const Rating = ({ rating }) => {
    const stars = [];
    for (let i = 0; i < 5; i++) {
        stars.push(
            <span className={emptyStar} key={i}>
                <svg viewBox="0 0 51 48" width="19" height="19">
                    <path
                        d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                        stroke="#002b57"
                        strokeWidth="2"
                        fill="none"
                    />
                </svg>
            </span>
        );
    }
    return (
        <div className={ratingContainer}>
            <div className={starsContainer}>
                <div className={filledStarsContainer}>
                    <StarRatings
                        rating={rating}
                        starDimension="20px"
                        starSpacing="1px"
                        starRatedColor="#002b57"
                        starEmptyColor="transparent"
                    />
                </div>
                <div className={emptyStarsContainer}>{stars}</div>
            </div>
        </div>
    );
};

Rating.propTypes = { rating: PropTypes.number };

export default Rating;
