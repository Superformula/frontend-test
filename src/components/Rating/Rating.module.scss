.ratingContainer {
	position: relative;
}
.filledStarsContainer {
	position: absolute;
	top: 0;
}
.emptyStarsContainer {
	position: absolute;
	top: 1px;
}
.emptyStar {
	width: 20px;
	height: 20px;
	padding: 1px;
	margin-right: 1px;
}
.starsContainer {
	display: inline-block;
}
