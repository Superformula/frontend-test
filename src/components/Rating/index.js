import React, { useMemo, memo } from 'react';
import PropTypes from 'prop-types';
import { ReactComponent as HalfStar } from 'assets/star-half.svg';
import { Star } from './Star';
import { Wrapper } from './Wrapper';

export const Rating = memo(({ rating, max, $xs, $lg }) => {
  const baseStars = new Array(max || 5).fill(null);
  const ratingStars = useMemo(
    () =>
      baseStars.map((_, index) => {
        const isFull = rating >= index + 1;
        const isHalf = Math.abs(index + 1 - rating) < 1;
        return <Star key={index} $fill={isFull} as={isHalf ? HalfStar : ''} />;
      }),
    [rating]
  );

  return <Wrapper {...{ $xs, $lg }}>{ratingStars}</Wrapper>;
});

Rating.displayName = 'Rating';

Rating.propTypes = {
  rating: PropTypes.number,
  max: PropTypes.number,
  $xs: PropTypes.bool,
  $lg: PropTypes.bool,
};

Rating.defaultProps = {
  rating: 0,
  max: 5,
  $xs: false,
  $lg: false,
};
