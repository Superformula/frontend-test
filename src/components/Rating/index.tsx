import React from "react";
import styled from "@emotion/styled";
import Star from "./Star";

export interface IProps {
  score: number;
}

// TODO make it scalable in size
export default function Rating(props: IProps): JSX.Element {
  // ensure score is between 0 and 100
  const score = Math.max(0, Math.min(100, props.score));
  const starCount = 5;
  const maxStarScore = 100 / starCount;
  const fullStarCount = Math.floor(score / maxStarScore);
  const scoreModulus = score % maxStarScore;
  const partialStarFullness = (scoreModulus / maxStarScore) * 100;

  return (
    <Wrapper>
      {[...Array(5).keys()].map((index) => {
        let fullness = 0;
        if (index < fullStarCount) {
          fullness = 100;
        }

        if (index === fullStarCount) {
          fullness = partialStarFullness;
        }

        return <Star key={index} fullness={fullness} />;
      })}
    </Wrapper>
  );
}

const Wrapper = styled.div`
  & > * {
    padding-right: 2px;
  }
`;
