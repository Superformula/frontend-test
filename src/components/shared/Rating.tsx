import * as React from "react";
import Icon from "./Icon/Icon";

interface Props {
  rating: number;
}

const Rating: React.FC<Props> = ({ rating }) => {
  const value = rating * 2;
  const stars = [];
  for (let i = 2; i <= 10; i += 2) {
    if (value >= i) {
      stars.push(<Icon name="starFill" />);
    } else if (value === i - 1) {
      stars.push(<Icon name="starHalf" />);
    } else {
      stars.push(<Icon name="starEmpty" />);
    }
  }

  return <>{stars}</>;
};

export default Rating;
