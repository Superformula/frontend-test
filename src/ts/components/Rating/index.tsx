// Imports
// -------

// Libraries
import React from 'react';
// Images
import emptyStar from '../../../img/emptyStar.svg';
import halfStar from '../../../img/halfStar.svg';
import fullStar from '../../../img/fullStar.svg';
// Styles
import style from './style.module.css';

// Internal
// --------

export interface RatingProps {
    value: 0 | 0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5,
    size?: 'small' | 'large'
}

const Rating: React.FC<RatingProps> = ({value, size = 'small'}) => {
    const getStar = (imgSrc: string, idx: number) => (
        <img src={ imgSrc } alt={ imgSrc } key={ idx } className={ style[size] }/>
    );

    //
    // Render

    const numFullStars = Math.floor(value);
    const numHalfStars = value - numFullStars === 0.5 ? 1 : 0;
    const numEmptyStars = Math.floor(5 - value);

    return (
        <div>
            { Array.from({length: numFullStars}).map((_, idx) => getStar(fullStar, idx)) }
            { Array.from({length: numHalfStars}).map((_, idx) => getStar(halfStar, idx)) }
            { Array.from({length: numEmptyStars}).map((_, idx) => getStar(emptyStar, idx)) }
        </div>
    );
}

// Exports
// -------

export default Rating;
