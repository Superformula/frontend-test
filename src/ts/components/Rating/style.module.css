.small {
    height: 19px;
}

.large {
    height: 30px;
}
