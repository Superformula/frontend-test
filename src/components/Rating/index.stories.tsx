import React from "react";
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from "@storybook/react/types-6-0";

import Rating, { IProps } from "./index";

export default {
  title: "Rating",
  component: Rating,
  args: {},
} as Meta;

const Template: Story<IProps> = (args) => <Rating {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  score: 50,
};
