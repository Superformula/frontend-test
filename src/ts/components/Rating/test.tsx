// Imports
// -------

// Libraries
import React from 'react';
import renderer from 'react-test-renderer';
// Components
import Rating from './index';

// Mocks
// -----

jest.mock("../../../img/emptyStar.svg", () => 'emptyStar.svg');
jest.mock("../../../img/halfStar.svg", () => 'halfStar.svg');
jest.mock("../../../img/fullStar.svg", () => 'fullStar.svg');

// Internal
// --------

const defaultProps = {
    value: 3
};

const getComponent = (props = {}) => renderer.create(
    <Rating {...defaultProps} {...props} />,
);

// Tests
// -----

describe('src/ts/components/Rating', () => {
    let component;

    describe('Rendering', () => {
        it('renders default correctly', () => {
            component = getComponent();
            expect(component).toMatchSnapshot();
        });
    });
});
