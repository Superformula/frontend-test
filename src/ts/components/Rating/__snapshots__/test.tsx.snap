// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/ts/components/Rating Rendering renders default correctly 1`] = `
<div>
  <img
    alt="fullStar.svg"
    className="small"
    src="fullStar.svg"
  />
  <img
    alt="fullStar.svg"
    className="small"
    src="fullStar.svg"
  />
  <img
    alt="fullStar.svg"
    className="small"
    src="fullStar.svg"
  />
  <img
    alt="emptyStar.svg"
    className="small"
    src="emptyStar.svg"
  />
  <img
    alt="emptyStar.svg"
    className="small"
    src="emptyStar.svg"
  />
</div>
`;
